### PR TITLE
Fix non-fixed types serialization

### PR DESCRIPTION
**Pull request by boguslavsky**

### CODE REVIEW

File: test/serialization.js [hunk 1]
```diff
@@ -57,7 +57,33 @@ describe('Serialize data into array of 8-bit integers', function() {
...
+        var User = Exonum.newType({
+            size: 12,
+            fields: {
+                name: {type: Exonum.FixedBuffer, size: 12, from: 0, to: 12}
+            }
+        });
+        var Transaction = Exonum.newType({
+            size: 16,
+            fields: {
+                to: {type: User, size: 8, from: 0, to: 8},
```
Comment (slowli) [inline test/serialization.js]: `User` is a fixed-length type (length is always 12), but it's serialized as a segment apparently.

File: src/types/serialization.js [hunk 2]
```diff
@@ -20,7 +20,7 @@ export function serialize(buffer, shift, data, type) {
                 if (!isFixed(fields[fieldName].type.fields)) {
                     return false;
                 }
-            } else if (fields[fieldName].type === String) {
+            } else if (fields[fieldName].type === String || fields[fieldName].type === FixedBuffer) {
```
Comment (slowli) [inline src/types/serialization.js]: Why though? `FixedBufer`s have fixed length by definition, as far as I understand.

File: src/types/serialization.js [hunk 3]
```diff
@@ -50,12 +50,12 @@ export function serialize(buffer, shift, data, type) {
...
             } else {
                 var end = buffer.length;
-                Uint32(end, buffer, from, from + 4);
+                Uint32(end - shift, buffer, from, from + 4);
                 buffer = serialize(buffer, end, fieldData, fieldType.type);
                 Uint32(buffer.length - end, buffer, from + 4, from + 8);
             }
         } else {
-            buffer = fieldType.type(fieldData, buffer, from, shift + fieldType.to);
+            buffer = fieldType.type(fieldData, buffer, from, shift + fieldType.to, buffer.length - shift);
```
Comment (slowli) [inline src/types/serialization.js]: What values can `fieldType.type` assume here? Only `String`?

File: src/types/serialization.js [hunk 4]
```diff
@@ -50,12 +50,12 @@ export function serialize(buffer, shift, data, type) {
                 buffer = serialize(buffer, from, fieldData, fieldType.type);
             } else {
                 var end = buffer.length;
-                Uint32(end, buffer, from, from + 4);
+                Uint32(end - shift, buffer, from, from + 4);
```
Comment (slowli) [inline src/types/serialization.js]: - This requires to update the documentation, doesn't it? Seems like a fundamental change in what segment pointers point to.
- How many unit tests cover this directly? Seems like too little.

File: test/serialization.js (hunk 1 above)
Comment (boguslavsky) [reply test/serialization.js]: In this PR I changed serialization according the rules that works in Exonum core right now. All nested types are serialized as segments without any review to their contents.

File: src/types/serialization.js (hunk 2 above)
Comment (boguslavsky) [reply src/types/serialization.js]: Thats because `FixedBuffer` is serialized as segment in Exonum core.

File: src/types/serialization.js (hunk 3 above)
Comment (boguslavsky) [reply src/types/serialization.js]: 1. `String`
2. `FixedBuffer`
3. entity of `Exonum.newType`

File: src/types/serialization.js (hunk 4 above)
Comment (boguslavsky) [reply src/types/serialization.js]: - Right, this changed should be reflected in documentation. We discussed it with @oudod.
- Now it is only one unit test for each case where segment pointers are used: 1) nested type 2) `String` 2) `FixedBuffer`

**Comment by boguslavsky** [comment]
@slowli PR is ready for review. Can you please check it again.

**Pull request merged**